test/jdk/java/lang/StringBuffer/IndexOf.java

```diff
@@ -27,162 +27,396 @@
+            + targetString.length() + ") String = "
+            + sourceBuffer.toString() + " len Buffer = " + sourceBuffer.toString().length());
+        System.err.println("  naive = " + naiveFind(sourceBuffer.toString(), targetString, 0) + ", IndexOf = "
+            + sourceBuffer.indexOf(targetString, 5));
+        failCount++;
+      }
+      if ((sourceBuffer.indexOf(targetString, 99) == index1) ||
+          (index1 != naiveFind(sourceBuffer.toString(), targetString, 0))) {
+        System.err.println("sourceBuffer.indexOf(targetString, 99) fragment '" + targetString + "' ("
+            + targetString.length() + ") String = "
+            + sourceBuffer.toString() + " len Buffer = " + sourceBuffer.toString().length());
+        System.err.println("  naive = " + naiveFind(sourceBuffer.toString(), targetString, 0) + ", IndexOf = "
+            + sourceBuffer.indexOf(targetString, 99));
+        failCount++;
+      }
+      if ((sourceBuffer.indexOf(emptyString, 99) != 99) ||
+          (99 != naiveFind(sourceBuffer.toString(), emptyString, 99))) {
+        System.err.println("sourceBuffer.indexOf(emptyString, 99) fragment '" + emptyString + "' ("
+            + emptyString.length() + ") String = "
+            + sourceBuffer.toString() + " len Buffer = " + sourceBuffer.toString().length());
+        System.err.println("  naive = " + naiveFind(sourceBuffer.toString(), emptyString, 99) + ", IndexOf = "
+            + sourceBuffer.indexOf(emptyString, 99));
+        failCount++;
+      }
+      if ((allAsBuffer.substring(1, 3).indexOf(allAsBuffer.substring(5, 12)) != -1) ||
+          (-1 != naiveFind(allAsBuffer.substring(1, 3).toString(), allAsBuffer.substring(5, 12), 0))) {
+        System.err.println("allAsBuffer.substring(1, 3).indexOf(allAsBuffer.substring(5, 12)) fragment '"
+            + allAsBuffer.substring(5, 12) + "' ("
+            + allAsBuffer.substring(5, 12).length() + ") String = "
+            + allAsBuffer.substring(1, 3) + " len Buffer = " + allAsBuffer.substring(1, 3).length());
+        System.err.println(
+            "  naive = " + naiveFind(allAsBuffer.substring(1, 3).toString(), allAsBuffer.substring(5, 12), 0)
+                + ", IndexOf = " + allAsBuffer.substring(1, 3).indexOf(allAsBuffer.substring(5, 12)));
+        failCount++;
+      }
+    }
 
-            if (sAnswer != sbAnswer)
-                failCount++;
+    report("Basic Test                   ", failCount);
+  }
+
+  // Note: it is possible although highly improbable that failCount will
+  // be > 0 even if everthing is working ok
+  private static void compareIndexOfLastIndexOf() {
+    int failCount = 0;
+    String sourceString;
+    StringBuffer sourceBuffer;
+    String targetString;
+
+    for (int i = 0; i < 10000; i++) {
+      do {
+        sourceString = generateTestString(99, 100);
+        sourceBuffer = new StringBuffer(sourceString);
+        targetString = generateTestString(10, 11);
```

Comment:
Generate a random int [0,1,2] for LL, UU, UL, pass that as parameter to generateTestString() to test the other paths. Same for other tests in this file using this pattern.

This test is specific to haystacklen=100, needlelen=10..  what about other haystack/needle sizes to exercise all the paths in the intrinsic assembler (i.e. haystack >=, <=32, needlelen ={1,2,3,4,5..32..}). Elsewhere already?